File: src/main/scala/us/zoom/data/dfence/providers/snowflake/matchers/PlaybookGrantMatcher.scala
package us.zoom.data.dfence.providers.snowflake.matchers

import org.slf4j.LoggerFactory
import us.zoom.data.dfence.providers.snowflake.models.{PlaybookGrant, SnowflakeGrant}

trait PlaybookGrantMatcher:
  def matchesPlaybookGrant(playbookGrant: PlaybookGrant): Boolean

object PlaybookGrantMatcher:
  def apply(snowflakeGrant: SnowflakeGrant): PlaybookGrantMatcher =
    PlaybookGrantMatcherLive(snowflakeGrant)

private class PlaybookGrantMatcherLive(snowflakeGrant: SnowflakeGrant) extends PlaybookGrantMatcher:

  private val log = LoggerFactory.getLogger(getClass)

  def matchesPlaybookGrant(playbookGrant: PlaybookGrant): Boolean =
    val matchers = SnowflakeGrantMatchers(playbookGrant)
    import matchers.*

    val result = snowflakeGrant match
      case SnowflakeGrant(GrantObjectType(), GrantPrivilege(), GrantType(), GrantObjectName()) => true
      case _                                                                                   => false

    if !result then
      log.info(
        "Playbook grant {} match failed for snowflake grant {}",
        playbookGrant,
        snowflakeGrant
      )
    else
      log.info(
        "Playbook grant {} matched snowflake grant {}",
        playbookGrant,
        snowflakeGrant
      )
    result

File: src/main/java/us/zoom/data/dfence/providers/snowflake/revoke/matchers/playbook/PlaybookGrantMatcher.java
package us.zoom.data.dfence.providers.snowflake.revoke.matchers.playbook;

import static us.zoom.data.dfence.providers.snowflake.revoke.matchers.SnowflakeGrantMatchers.*;

import java.util.function.BiPredicate;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import us.zoom.data.dfence.providers.snowflake.revoke.models.PlaybookGrant;
import us.zoom.data.dfence.providers.snowflake.revoke.models.SnowflakeGrant;

@Slf4j
@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class PlaybookGrantMatcher {

  /**
   * Creates a predicate that matches a SnowflakeGrant against a PlaybookGrant. Checks object type,
   * privilege, grant type, and object name.
   */
  public static BiPredicate<PlaybookGrant, SnowflakeGrant> matchGrantAgainstPlaybook() {
    return (playbookGrant, snowflakeGrant) ->
        matchesGrantObjectType(snowflakeGrant.snowflakeObjectType())
            .and(matchesGrantPrivilege(snowflakeGrant.privilege()))
            .and(matchesGrantType(snowflakeGrant.grantType()))
            .and(matchesGrantName(snowflakeGrant.name(), snowflakeGrant.snowflakeObjectType()))
            .test(playbookGrant);
  }
}
